### PR TITLE
Fixes for TaskKillService

### PR DESCRIPTION
**Pull request by meichstedt**
**Please don't squash merge**

Cherry picking 2 commits from `feature/pods` to master so they can be cherry-picked onto 1.3 afterwards:

### Introducing UnknownTaskTerminated event
Until now, no event was dispatched on the event stream when the related task was terminal and unknown, thus making the TaskKillService unable to track the progress of these kill requests. Promises created by TaskKillProgressActors were never completed, leading to a memory leak. With this change, an UnknownTaskTerminated event will be dispatched on the event bus when a terminal status update for an unknown task is received. This event is also visible for subscribers to the event stream. The TaskKillService handles this event and correctly recognizes it as a kill success.

Cherry-picked from feature/pods (was: Adjust the KillService to handle instances instead of tasks (#4414))

### Fix KillService behavior: retry forever and don't spawn progress actors for unknown tasks

* killUnknownTask now has a unit return type (no one is interested in the result)
* Do not create KillProgressActors for unknown tasks
* remove maxRetries; if no terminal task status is received, the killService will retry until the task is confirmed to be terminal

### CODE REVIEW

File: src/main/scala/mesosphere/marathon/core/task/termination/impl/TaskKillServiceActor.scala [hunk 1]
```diff
@@ -114,21 +116,25 @@ private[impl] class TaskKillServiceActor(
...
+      task.isGone || task.isUnknown || task.isDropped || task.isUnreachable
+    }
 
     if (taskIsLost) {
       log.warning("Expunging lost {} from state because it should be killed", taskId)
       // we will eventually be notified of a taskStatusUpdate after the task has been expunged
       stateOpProcessor.process(TaskStateOp.ForceExpunge(taskId))
     } else {
-      val knownOrNot = if (maybeTask.isDefined) "known" else "unknown"
+      val knownOrNot = if (toKill.maybeTask.isDefined) "known" else "unknown"
```
Comment (jeschkies) [inline src/main/scala/mesosphere/marathon/core/task/termination/impl/TaskKillServiceActor.scala]: `toKill.fold("unknown")(task => "known")` is an alternative :)

File: src/main/scala/mesosphere/marathon/core/task/update/impl/TaskStatusUpdateProcessorImpl.scala [hunk 2]
```diff
@@ -72,18 +82,28 @@ class TaskStatusUpdateProcessorImpl @Inject() (
...
+    MarathonTaskStatus.Gone,
+    MarathonTaskStatus.Dropped,
+    MarathonTaskStatus.Unknown
   )
   // It doesn't make sense to kill an unknown task if it is in a terminal or killing state
   // We'd only get another update for the same task
-  private def killWhenUnknown(status: MesosProtos.TaskStatus): Boolean = {
-    !ignoreWhenUnknown.contains(status.getState)
+  private def killWhenUnknown(status: MarathonTaskStatus): Boolean = {
+    !ignoreWhenUnknown.contains(status)
```
Comment (jeschkies) [inline src/main/scala/mesosphere/marathon/core/task/update/impl/TaskStatusUpdateProcessorImpl.scala]: Call it terminal state then

```scala
val terminalStates = Set (...)

killWhenNonTerminal(status: MarathonTaskStatus): Boolean = !terminalStates.contain(status)
```

Comment (jeschkies) [inline src/main/scala/mesosphere/marathon/core/task/update/impl/TaskStatusUpdateProcessorImpl.scala]: Ah, the unknown is an unknown task. Hm, then I'd say `killWhenUnknownTask`. However, see above.

File: src/main/scala/mesosphere/marathon/core/task/update/impl/TaskStatusUpdateProcessorImpl.scala [hunk 3]
```diff
@@ -43,13 +47,19 @@ class TaskStatusUpdateProcessorImpl @Inject() (
...
         val taskStateOp = TaskStateOp.MesosUpdate(task, status, now)
         stateOpProcessor.process(taskStateOp).flatMap(_ => acknowledge(status))
 
-      case None if killWhenUnknown(status) =>
+      case None if terminalUnknown(marathonTaskStatus) =>
+        log.warn("Received terminal status update for unknown {}", taskId)
+        eventStream.publish(UnknownTaskTerminated(taskId, taskId.runSpecId, marathonTaskStatus))
+        acknowledge(status)
+
+      case None if killWhenUnknown(marathonTaskStatus) =>
```
Comment (jeschkies) [inline src/main/scala/mesosphere/marathon/core/task/update/impl/TaskStatusUpdateProcessorImpl.scala]: I'd just say `case None if nonTerminalState(marathonTaskStatus)`.

File: src/main/scala/mesosphere/marathon/core/task/update/impl/TaskStatusUpdateProcessorImpl.scala [hunk 4]
```diff
@@ -72,18 +82,28 @@ class TaskStatusUpdateProcessorImpl @Inject() (
...
   lazy val name = Names.named(getClass.getSimpleName)
 
+  /** Matches all states that are considered terminal for an unknown task */
+  private[impl] def terminalUnknown(status: MarathonTaskStatus): Boolean = status match {
+    case t: MarathonTaskStatus.Terminal => true
+    case MarathonTaskStatus.Unreachable => true
+    case _ => false
+  }
+
   private[this] val ignoreWhenUnknown = Set(
```
Comment (jeschkies) [inline src/main/scala/mesosphere/marathon/core/task/update/impl/TaskStatusUpdateProcessorImpl.scala]: I'd actually move this as `terminalStates` to `mesosphere.marathon.core.task.state.MarathonTaskStatus` since this will be a common thing and should be defined in one position.

**Comment by aquamatthias** [comment]
retest this please

File: src/main/scala/mesosphere/marathon/core/task/update/impl/TaskStatusUpdateProcessorImpl.scala (hunk 2 above)
Comment (unterstein) [inline src/main/scala/mesosphere/marathon/core/task/update/impl/TaskStatusUpdateProcessorImpl.scala]: another possible naming could be `killWhenTaskUnknown` ;)

File: src/main/scala/mesosphere/marathon/core/task/update/impl/TaskStatusUpdateProcessorImpl.scala (hunk 3 above)
Comment (meichstedt) [inline src/main/scala/mesosphere/marathon/core/task/update/impl/TaskStatusUpdateProcessorImpl.scala]: Which wouldn't be 100% correct – `Unreachable` is a non-terminal state, but killing a task that's unreachable will just lead to another status update that says it's `Unreachable`. There's a reason for being so explicit, but it's obviously not clear, so I'll document that.

File: src/main/scala/mesosphere/marathon/core/task/update/impl/TaskStatusUpdateProcessorImpl.scala (hunk 2 above)
Comment (meichstedt) [inline src/main/scala/mesosphere/marathon/core/task/update/impl/TaskStatusUpdateProcessorImpl.scala]: I didn't address these comments here so we can proceed with cherry-picking into 1.3. I agree we should clarify naming and use-case; let's do that in a separate PR.

File: src/main/scala/mesosphere/marathon/core/task/update/impl/TaskStatusUpdateProcessorImpl.scala (hunk 3 above)
Comment (jeschkies) [inline src/main/scala/mesosphere/marathon/core/task/update/impl/TaskStatusUpdateProcessorImpl.scala]: Hm, why would killing an `Unreachable` task trigger a status update? I thought Mesos would not do much with it.

Comment (meichstedt) [inline src/main/scala/mesosphere/marathon/core/task/update/impl/TaskStatusUpdateProcessorImpl.scala]: I am not sure how this behavior changes with Mesos 1.1 or future versions; until now, when killing a task that was scheduled on an agent that is currently not registered, Mesos sent a statusUpdate declaring the task is LOST (which makes sense because the master cannot forward the request to the agent).
